src/moonlink_datafusion/src/connection_pool.rs

```diff
@@ -0,0 +1,361 @@
+use std::{
+    collections::HashMap,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        LazyLock,
+    },
+    time::{Duration, Instant},
+};
+use tokio::{net::UnixStream, sync::Mutex, time::sleep};
+
+struct PooledEntry {
+    stream: UnixStream,
+    last_used: Instant,
+}
+
+static POOL: LazyLock<Mutex<HashMap<String, Mutex<Vec<PooledEntry>>>>> =
+    LazyLock::new(|| Mutex::new(HashMap::new()));
+
+const MAX_PER_URI: usize = 30;
+
+static IDLE_TIMEOUT_MS: AtomicU64 = AtomicU64::new(30_000);
+
+pub fn idle_timeout() -> Duration {
+    Duration::from_millis(IDLE_TIMEOUT_MS.load(Ordering::Relaxed))
+}
+
+#[derive(Debug)]
+pub struct PooledStream {
+    pub uri: String,
+    pub stream: Option<UnixStream>,
+}
+
+impl PooledStream {
+    pub fn new(uri: String, stream: UnixStream) -> Self {
+        Self {
+            uri,
+            stream: Some(stream),
+        }
+    }
+    pub fn stream_mut(&mut self) -> &mut UnixStream {
+        self.stream.as_mut().unwrap()
+    }
+}
+
+impl Drop for PooledStream {
+    fn drop(&mut self) {
+        if let Some(stream) = self.stream.take() {
+            let uri = self.uri.clone();
+            tokio::spawn(async move {
+                let mut pool = POOL.lock().await;
+                let mut pool_vec = pool.entry(uri).or_default().lock().await;
+
+                if pool_vec.len() < MAX_PER_URI {
+                    pool_vec.push(PooledEntry {
+                        stream,
+                        last_used: Instant::now(),
+                    });
+                }
+            });
+        }
+    }
+}
+
+pub(crate) async fn get_stream(uri: &str) -> crate::Result<PooledStream> {
+    {
+        let mut pool = POOL.lock().await;
+        if let Some(mutex_vec) = pool.get_mut(uri) {
+            let mut vec = mutex_vec.lock().await;
+            // Remove expired streams
+            vec.retain(|entry| entry.last_used.elapsed() <= idle_timeout());
+
+            while let Some(entry) = vec.pop() {
+                if is_connection_alive(&entry.stream).await {
```

Comment:
please don't put IO operations inside of critical section
for implementation easiness, we could return the stream directly